application/src/main/java/org/thingsboard/server/service/entitiy/queue/DefaultTbQueueService.java

```diff
@@ -0,0 +1,276 @@
+    private final TbQueueAdmin tbQueueAdmin;
+    private final DeviceProfileService deviceProfileService;
+
+    @Override
+    public Queue saveQueue(Queue queue) {
+        boolean create = queue.getId() == null;
+        Queue oldQueue;
+
+        if (create) {
+            oldQueue = null;
+        } else {
+            oldQueue = queueService.findQueueById(queue.getTenantId(), queue.getId());
+        }
+
+        //TODO: add checkNotNull
+        Queue savedQueue = queueService.saveQueue(queue);
+
+        if (create) {
+            onQueueCreated(savedQueue);
+        } else {
+            onQueueUpdated(savedQueue, oldQueue);
+        }
+
+        return savedQueue;
+    }
+
+    @Override
+    public void deleteQueue(TenantId tenantId, QueueId queueId) {
+        Queue queue = queueService.findQueueById(tenantId, queueId);
+        queueService.deleteQueue(tenantId, queueId);
+        onQueueDeleted(tenantId, queue);
+    }
+
+    @Override
+    public void deleteQueueByQueueName(TenantId tenantId, String queueName) {
+        Queue queue = queueService.findQueueByTenantIdAndNameInternal(tenantId, queueName);
+        queueService.deleteQueue(tenantId, queue.getId());
+        onQueueDeleted(tenantId, queue);
+    }
+
+    private void onQueueCreated(Queue queue) {
+        if (tbQueueAdmin != null) {
+            for (int i = 0; i < queue.getPartitions(); i++) {
+                tbQueueAdmin.createTopicIfNotExists(
+                        new TopicPartitionInfo(queue.getTopic(), queue.getTenantId(), i, false).getFullTopicName());
+            }
+        }
+
+        if (tbClusterService != null) {
+            tbClusterService.onQueueChange(queue);
+        }
+    }
+
+    private void onQueueUpdated(Queue queue, Queue oldQueue) {
+        int oldPartitions = oldQueue.getPartitions();
+        int currentPartitions = queue.getPartitions();
+
+        if (currentPartitions != oldPartitions && tbQueueAdmin != null) {
+            if (currentPartitions > oldPartitions) {
+                log.info("Added [{}] new partitions to [{}] queue", currentPartitions - oldPartitions, queue.getName());
+                for (int i = oldPartitions; i < currentPartitions; i++) {
+                    tbQueueAdmin.createTopicIfNotExists(
+                            new TopicPartitionInfo(queue.getTopic(), queue.getTenantId(), i, false).getFullTopicName());
+                }
+                if (tbClusterService != null) {
+                    tbClusterService.onQueueChange(queue);
+                }
+            } else {
+                log.info("Removed [{}] partitions from [{}] queue", oldPartitions - currentPartitions, queue.getName());
+                if (tbClusterService != null) {
+                    tbClusterService.onQueueChange(queue);
+                }
+                await();
+                for (int i = currentPartitions; i < oldPartitions; i++) {
+                    tbQueueAdmin.deleteTopic(
```

Comment:
add logs on info or warn level about destructive or any other change operations. This will help to debug race-conditions in cluster mode (if any)